applications/StructuralMechanicsApplication/custom_processes/spr_error_process.cpp

```diff
@@ -0,0 +1,531 @@
+    ElementsArrayType& elements_array = mThisModelPart.Elements();
+    const int num_elem = static_cast<int>(elements_array.size());
+
+    // Compute the error estimate per element
+    double error_overall= 0.0;
+    double energy_norm_overall = 0.0;
+    #pragma omp parallel for reduction(+:error_overall, energy_norm_overall)
+    for(int i_elem = 0; i_elem < num_elem; ++i_elem){
+        auto it_elem = elements_array.begin() + i_elem;
+
+        std::vector<double> error_integration_point;
+        const auto& process_info = mThisModelPart.GetProcessInfo();
+        it_elem->GetValueOnIntegrationPoints(ERROR_INTEGRATION_POINT, error_integration_point, process_info);
+
+        KRATOS_INFO_IF("SPRErrorProcess", mEchoLevel > 2) << "Error GP:" << error_integration_point << std::endl;
+        
+        // We compute the error overall 
+        double error_energy_norm = 0.0;
+        for(IndexType i = 0;i < error_integration_point.size();++i)
+            error_energy_norm += error_integration_point[i];
+        error_overall += error_energy_norm;
+        error_energy_norm = std::sqrt(error_energy_norm);
+        it_elem->SetValue(ELEMENT_ERROR, error_energy_norm);
+
+
+        // We compute now the energy norm
+        std::vector<double> strain_energy;
+        it_elem->GetValueOnIntegrationPoints(STRAIN_ENERGY, strain_energy, process_info);
+
+        double energy_norm = 0.0;
+        for(IndexType i = 0;i < strain_energy.size(); ++i)
+            energy_norm += 2.0 * strain_energy[i];
+        energy_norm_overall += energy_norm;
+        energy_norm= std::sqrt(energy_norm);
+
+        KRATOS_INFO_IF("SPRErrorProcess", mEchoLevel > 2) << "Element Id:" << it_elem->Id() << ". Element error: " << error_energy_norm << ". Energy norm: " << energy_norm << std::endl;
+    }
+
+    rErrorOverall = std::sqrt(error_overall);
+    rEnergyNormOverall = std::sqrt(energy_norm_overall);
+    double error_percentage = rErrorOverall/std::sqrt((std::pow(rErrorOverall, 2) + std::pow(rEnergyNormOverall, 2)));
+
+    KRATOS_INFO_IF("SPRErrorProcess", mEchoLevel > 1)
+        << "Overall error norm: " << rErrorOverall << std::endl
+        << "Overall energy norm: "<< rEnergyNormOverall << std::endl
+        << "Error in percent: " << error_percentage << std::endl;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRErrorProcess<TDim>::CalculatePatch(
+    NodeItType itNode,
+    NodeItType itPatchNode,
+    SizeType NeighbourSize,
+    Vector& rSigmaRecovered
+    )
+{
+    // Determine if contact BC has to be regarded
+    // We take the geometry GP from the core
+    const double tolerance = std::numeric_limits<double>::epsilon();
+    const bool regard_contact = itNode->Has(CONTACT_PRESSURE) ? (std::abs(itNode->GetValue(CONTACT_PRESSURE)) > tolerance ? true : false) : false;
```

Comment:
this is specific to contact. We should have a standard ZZ estimator and then one for contact.

maybe we could make the function CalculatePatch virtual and derive one process from the other?
or maybe we could set at configuration time if contact is to be considered or not? (an option set as false by default)

Reply:
btw, what type of sigma are u using?

PK2_STRESS?
CAUCHY_STRESS?

i think this should be selected by the user (with appropriate defaults)

Reply:
CAUCHY_STRESS, this is the one supposed to be used, not any other

Reply:
I don't agree to separate in contact, it is automatically detected, and I don't think it does harm

Reply:
PK2_STRESS could also be recovered...or any other stress measure

it would simplify the code to separate it. 
also if you don't you are paying a performance price

Reply:
OK, already doing many changes, but I push all together (think that there are 3 parallel PR here, and I need to check I don't break anything)